### DISCUSSION TITLE
ONNX inference without mmdet.apis

### DISCUSSION THREAD

**Discussion by tomyot** [open-mmlab/mmdeploy#2614] [General]
I am running into some issues running inference using onnxruntime.
After training a mask-rcnn model, I created an ONNX export using mmdeploy with the following flags:

```
python ./tools/deploy.py    
.../mmdeploy/configs/mmdet/detection/detection_onnxruntime_dynamic.py    .../mmdetection/configs/mask_rcnn/mask-rcnn_r50_fpn_1x_coco.py   
.../epoch_10.pth"  
/home/data/im00001.tif  
--work-dir work-dir  
--show   
--device cuda:0
```

It is unclear to me whether the preprocessing is baked into the exported model or not.

I have tried the following preprocessing steps:

```
import torch
import os
import cv2
import onnxruntime as ort

class Preprocessor(torch.nn.Module):
    def __init__(self, norm_mean, norm_stdev):
        super().__init__()
        norm_mean = torch.asarray(norm_mean).reshape(-1, 1, 1).cuda()
        norm_stdev = torch.asarray(norm_stdev).reshape(-1, 1, 1).cuda()
        self.register_buffer('norm_mean', norm_mean)
        self.register_buffer('norm_stdev', norm_stdev)

    def forward(self, image):
        image = image.to(torch.float32)
        image = torch.stack([image, image, image], dim=0)
        image = (image - self.norm_mean) / self.norm_stdev
        image = image.unsqueeze(0)  
        return image
    

norm_mean = [103.530, 116.280, 123.675]
norm_mean = list(reversed(norm_mean))

preprocessor_config = {'norm_mean': norm_mean,  'norm_stdev': [1.0, 1.0, 1.0]}
models_folder = os.path.expanduser('~/models')
input_model_path = os.path.join(models_folder, 'export_mask_rcnn.onnx')
im_path = os.path.join(models_folder, 'test_image.tif')

img = cv2.imread(im_path, cv2.IMREAD_GRAYSCALE)
img = torch.asarray(img).cuda()

preproc = Preprocessor(**preprocessor_config)
img_f = preproc(img)
img_f = img_f.cpu().numpy()

providers = ['CUDAExecutionProvider', 'CPUExecutionProvider']
ort_sess = ort.InferenceSession(input_model_path, providers=providers)
outputs = ort_sess.run(['dets', 'labels'], {'input': img_f})
```

but I am getting different results compared to inference with mmdet.apis.inference_detector.

What is the expected input to the ONNX model? 

Thanks
